### ISSUE TITLE
How to produce events. Examples.

### ISSUE THREAD

**Issue by voland62** [funkia/purescript-hareactive#1]
It would be nice to have couple simple examples:
1. How to produce Stream from keyboard events. (Is there some "streamFromEvent" from hareactive counterpart?)
2. How to "subscribe" Stream to do some Effect.

Thx.

**Comment by paldepind** [comment]
Hello voland62. Yes, you are definitely right. Currently there are not way to do this in purescript-hareactive. My initial focus was to create bindings for the parts of Hareactive that are used when using [purescript-turbine](https://github.com/funkia/purescript-turbine).

I think purescript-hareactive should include something like a `Hareactive.Effects` module or something similar which would contain the functions for creating streams and behaviors and for "consuming" them and doing effects with them. What do you think about that?

**Comment by voland62** [comment]
Hello.
  Sorry, only now can continue explore hareactive...
Well, I'm  trying to work with scanS now, and I'm catching some runtime exception:
```javascript
const hare = require("@funkia/hareactive");

exports.down = function (){
    return hare.streamFromEvent(document, "keydown").map(function (e) {return e.key;});
};
```
```purescript
foreign import down ∷ Effect (E String)
count ∷ Effect (E Int)
count = down >>= scanS (\_ p → p + 1) 0 >>> switchStream >>> pure

main ∷ Effect Unit
main = do
  stream ∷ E Int ← count  
  let now = runStreamEffect (logShow <$> stream) 
  _ ← runNow now
  pure unit
```

And an exception:
Uncaught TypeError: Cannot read property 'addListener' of undefined
    at SwitchBehaviorStream.activate (stream.js:176)
    at SwitchBehaviorStream.addListener (common.js:36)
    at MapStream.activate (common.js:65)
    at MapStream.addListener (common.js:36)
    at new PerformCbStream (stream.js:391)
    at mapCbStream (stream.js:404)
    at PerformNow.perform [as cb] (now.js:113)
    at PerformNow.run (now.js:83)
    at exports._runNow (foreign.js:74)
    at foreign.js:66

It feels like it is looking for "initial stream..  but not sure..

**Comment by paldepind** [comment]
That looks like a bug in Hareactive.

However, calling `switchStream` on the result from `scanS` doesn't really make sense (which is also related to why it doesn't work). What you probably want is to use [`sample`](https://pursuit.purescript.org/packages/purescript-hareactive/0.0.5/docs/Hareactive.Combinators#v:sample) on the result from `scanS`.

Reading the section [How stateful behaviors work](https://github.com/funkia/hareactive#how-stateful-behaviors-work) may help you understand why `switchStream` is not the right function.

Basically, `scanS` returns a behavior which at any moment contains a stream that begins accumulating state in that moment. This means that the behavior changes to a new stream all the time. Hence, when calling `switchStream` on that behavior you'd just get a stream that would continuously switch to a brand new stream without any accumulated state.

Let me know if I should explain in more detail.

**Comment by paldepind** [comment]
Hi again @voland62.

> How to produce Stream from keyboard events. (Is there some "streamFromEvent" from hareactive counterpart?)

Yes. [streamFromEvent](https://pursuit.purescript.org/packages/purescript-hareactive/0.1.1/docs/Hareactive.DOM#v:streamFromEvent) has just been added in the new `Hareactive.DOM` module :smile: 

Here is some code which, I think, does what you want. A few names have changed slightly in Hareactive since you posted the original code.

```purescript
module Main where

import Prelude

import Effect (Effect)
import Effect.Class (liftEffect)
import Hareactive.Combinators (logS, runNow, scan)
import Hareactive.DOM (keyDown)
import Hareactive.Types (Now)

frpMain :: Now Unit
frpMain = do
  count <- scan (\_ p -> p + 1) 0 keyDown
  liftEffect $ logS "count:" (show <$> count)
  pure unit

main :: Effect Unit
main = runNow frpMain
```

I'll close the issue. Please reopen if you have further issues.
